### DISCUSSION TITLE
Fitview using only the width of the diagram

### DISCUSSION THREAD

**Discussion by KevinDaGame** [xyflow/xyflow#3522] [Q&A] [answered]
Hello! I've got a question I couldn't find the answer to. 
I'm working on a diagram view as a part of a dashboard for a client. This diagram will show multiple steps, and their status. The diagram is vertical, and has to fit in a slim (350px) area. In addition this region is restricted in heigh (600px). The diagram has panning disabled, and should only pan using the vertical scrollwheel. 

Therefore, the diagram should fit in such a way that the nodes are clearly readable, and no side panning is required.
However, fitView will result in the diagram being zoomed out to a point where nodes are not readable. What I would like is to be able to fit, but to ignore the vertical aspect,

I've made a simplified example to illustrate the issue.

## Image 1: The current situation where the nodes are unreadable
![image](https://github.com/wbkd/react-flow/assets/65958288/47109d6d-7ef3-446c-93dd-710ba93910c1)

## Image 2: What it should look like. The fit will only consider the with aspect
![image](https://github.com/wbkd/react-flow/assets/65958288/7176b5c7-ce6e-4f35-8127-dfdde6aecafb)

## My test project
[![Edit green-violet-7m8l3x](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/green-violet-7m8l3x?fontsize=14&hidenavigation=1&theme=dark)

How can I the fit consider only the width?

**Comment by KevinDaGame** [comment]
Hello, I am still looking for a solution for this problem

**Comment by bcakmakoglu** [comment, marked as answer]
Alright so for the first part about limiting the area in which you can pan, there is a prop called `translateExtent` that you can use.

It looks like this `[[x-from, y-from], [x-to, y-to]]`.

In your case, you would probably want to try to calculate the extent based on the positions your nodes have, so something like this

```ts
  useEffect(() => {
    const padding = 50;

    const lowestX = Math.min(...nodes.map((n) => n.position.x));
    const lowestY = Math.min(...nodes.map((n) => n.position.y));

    const highestX = Math.max(...nodes.map((n) => n.position.x));
    const highestY = Math.max(...nodes.map((n) => n.position.y));

    setTranslateExtent([
      [lowestX - padding, lowestY - padding],
	  // added 150 to account for the nodes width when calculating the max extent, replace with whatever you want
      [highestX + 150 + padding, highestY + padding * 2],
    ]);
  }, [nodes]);
```

For the `fitView` issue you could instead use `setViewport` and do something like this (this is just an example though)

```ts
onInit={(instance) => {
  const lowestX = Math.min(...nodes.map((n) => n.position.x));
  const lowestY = Math.min(...nodes.map((n) => n.position.y));

  const highestX = Math.max(...nodes.map((n) => n.position.x));
  const highestY = Math.max(...nodes.map((n) => n.position.y));

  instance.setViewport({
    x: (highestX - lowestX) / 2,
    y: 50,
    zoom: 0.75,
  });
}}
```

[Here's an example sandbox.](https://codesandbox.io/p/sandbox/serene-river-9zhlx3?file=%2Fsrc%2FApp.tsx%3A206%2C13-218%2C15)

**Reply by KevinDaGame** [reply]
Thank you for the response! I will look into implementing this. Regarding the second answer, our diagrams have a variable width (some might only be 1 node wide, while others might be three or more nodes wide. With the approach you outlined, would I have to calculate the zoom percentage myself, or is there a built in method for this?

**Reply by KevinDaGame** [reply]
Thank you! I've managed to get a working solution after fiddling with the numbers for a few hours. In the end, this is what worked for me:
```tsx
 const padding = 50;
        const lowestX = Math.min(...nodes.map((n) => n.position.x));
        const lowestY = Math.min(...nodes.map((n) => n.position.y));
        const highestXWithWidth = Math.max(...nodes.map((n) => n.position.x + (n.width ?? 0)));
        const diagramWidth = ref.current!.offsetWidth;

        const zoom = diagramWidth / (highestXWithWidth - lowestX - padding);
        const x = -lowestX * zoom;
        const y = -lowestY * zoom;

        console.log("Setting viewport to " + x + ", " + y + ", " + zoom);
        setViewport({
            x: x,
            y: y,
            zoom: zoom,
        });
```
